### PR TITLE
Fix bugs with Accuracy Aware

### PR DESCRIPTION
**Pull request by kshpv**
### Changes

Fix bugs with JSON schema, add a correct validation of Accuracy Aware config. 
Fix bug with non translating param validate_every_n_epochs from config. 
Add logging of validation metrics and final compression rate in adaptive compression level training pipeline. 
Improve documentation. 
Add typehints, add logging of saving checkpoint.
Update AA config of resnet-18 with new improved metrics.

### Reason for changes

Improving Accuracy Aware

### Related tickets

69079, 68302

### Tests

The current tests were changed accordingly.

### CODE REVIEW

**Comment by kshpv** [comment]
@MaximProshin, @AlexKoff88 please take a look at the new documentation section

**Comment by kshpv** [comment]
Jenkins please retry a build

File: nncf/common/accuracy_aware_training/runner.py [hunk 1]
```diff
@@ -142,13 +143,11 @@ def create_training_loop(self) -> TrainingRunner:
 
 class EarlyExitTrainingRunnerCreator(TrainingRunnerCreator):
     def __init__(self, accuracy_aware_training_params, compression_controller,
-                 lr_updates_needed, verbose, validate_every_n_epochs,
-                 dump_checkpoints):
+                 lr_updates_needed, verbose, dump_checkpoints):
```
Comment (vshampor) [inline nncf/common/accuracy_aware_training/runner.py:146]: Since you touched these, I'll ask you to set up type hints here and a docstring in this PR as well, please.

File: nncf/common/accuracy_aware_training/runner.py [hunk 2]
```diff
@@ -161,28 +160,24 @@ def create_training_loop(self) -> TrainingRunner:
...
+                                                 self.verbose, self.dump_checkpoints)
         raise RuntimeError('Got an unsupported value of nncf_backend')
 
 
 class AdaptiveCompressionLevelTrainingRunnerCreator(TrainingRunnerCreator):
     def __init__(self, accuracy_aware_training_params, compression_controller,
                  lr_updates_needed, verbose, minimal_compression_rate,
-                 maximal_compression_rate, validate_every_n_epochs,
-                 dump_checkpoints):
+                 maximal_compression_rate, dump_checkpoints):
```
Comment (vshampor) [inline nncf/common/accuracy_aware_training/runner.py:174]: Same.

**Review by vshampor — CHANGES_REQUESTED** [review]

File: nncf/common/accuracy_aware_training/runner.py (hunk 1 above)
Comment (kshpv) [reply nncf/common/accuracy_aware_training/runner.py:149]: Added small typehint

File: nncf/common/accuracy_aware_training/runner.py (hunk 2 above)
Comment (kshpv) [reply nncf/common/accuracy_aware_training/runner.py:180]: Added small typehint

Comment (kshpv) [reply nncf/common/accuracy_aware_training/runner.py:180]: @vshampor Are these typehints quite good?

**Review by vshampor — APPROVED** [review]

File: nncf/common/accuracy_aware_training/runner.py [hunk 3]
```diff
@@ -161,28 +176,29 @@ def create_training_loop(self) -> TrainingRunner:
...
-                                                 self.verbose, self.validate_every_n_epochs,
-                                                 self.dump_checkpoints)
+                                                 self.verbose, self.dump_checkpoints)
         raise RuntimeError('Got an unsupported value of nncf_backend')
 
 
 class AdaptiveCompressionLevelTrainingRunnerCreator(TrainingRunnerCreator):
-    def __init__(self, accuracy_aware_training_params, compression_controller,
-                 lr_updates_needed, verbose, minimal_compression_rate,
-                 maximal_compression_rate, validate_every_n_epochs,
```
Comment (AlexKoff88) [inline nncf/common/accuracy_aware_training/runner.py:177]: Why did you remove `validate_every_n_epochs`?

Comment (kshpv) [reply nncf/common/accuracy_aware_training/runner.py:177]: I placed it in `accuracy_aware_training_params` as other parameters from the config file

File: docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md [hunk 4]
```diff
@@ -1,6 +1,17 @@
...
+The compression level search can be only done for a single compression algorithm in the pipeline (i.e. several compression algorithms could be applied to the model and the search is going to be performed for a single one); currently supported algorithms for compression rate search are **magnitude sparsity** and **filter pruning**. 
+The exact compression algorithm for which the search is done is determined from `"compression"` config section. Below is an example of a filter pruning configuration with added `"accuracy_aware_training"` parameters. The parameters to be set by the user in this config section are:
```
Comment (AlexKoff88) [inline docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md:4]: "Below is an example of a filter pruning configuration with added `"accuracy_aware_training"` parameters." - should be removed or moved to right before the example.

File: docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md [hunk 5]
```diff
@@ -1,6 +1,17 @@
...
+The compression level search can be only done for a single compression algorithm in the pipeline (i.e. several compression algorithms could be applied to the model and the search is going to be performed for a single one); currently supported algorithms for compression rate search are **magnitude sparsity** and **filter pruning**.
```
Comment (AlexKoff88) [inline docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md:3]: Can you provide few words in the beginning what is it aimed for? It should be clear to the user what he/she came across. My suggestion is:
"Adaptive Compression Level training loop is the meta-algorithm that allows searching for the compression level of the underneath compression methods while staying within the range of the user-defined maximum accuracy degradation".

File: docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md [hunk 6]
```diff
@@ -1,6 +1,17 @@
 # Adaptive Compression Level training loop in NNCF
```
Comment (MaximProshin) [inline docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md:1]: It is referenced from Usage.md but the reference there is not in the right place. I propose to add the reference the first time "Adaptive Compression Level Training" is mentioned. The same as for "Early Exit Training" can you fix that?

File: docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md [hunk 7]
```diff
@@ -1,6 +1,17 @@
...
+The compression level search can be only done for a single compression algorithm in the pipeline (i.e. several compression algorithms could be applied to the model and the search is going to be performed for a single one); currently supported algorithms for compression rate search are magnitude sparsity and filter pruning.
```
Comment (MaximProshin) [inline docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md]: Let's use capital letters for magnitude sparsity and filter pruning.

File: docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md [hunk 8]
```diff
@@ -1,6 +1,17 @@
...
+The compression level search can be only done for a single compression algorithm in the pipeline (i.e. several compression algorithms could be applied to the model and the search is going to be performed for a single one); currently supported algorithms for compression rate search are magnitude sparsity and filter pruning. 
+The exact compression algorithm for which the search is done is determined from `"compression"` config section. Below is an example of a filter pruning configuration with added `"accuracy_aware_training"` parameters. The parameters to be set by the user in this config section are:
```
Comment (MaximProshin) [inline docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md:4]: The exact compression algorithm for which the search is done is determined from "compression" config section
=>
The exact compression algorithm is determined in"compression" config section

File: docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md [hunk 9]
```diff
@@ -1,6 +1,17 @@
...
+The compression level search can be only done for a single compression algorithm in the pipeline (i.e. several compression algorithms could be applied to the model and the search is going to be performed for a single one); currently supported algorithms for compression rate search are magnitude sparsity and filter pruning. 
+The exact compression algorithm for which the search is done is determined from `"compression"` config section. Below is an example of a filter pruning configuration with added `"accuracy_aware_training"` parameters. The parameters to be set by the user in this config section are: 
+1) `maximal_relative_accuracy_degradation` or `maximal_absolute_accuracy_degradation` - the maximal allowed accuracy metric drop relative to the original model metrics (in percent) or the maximal allowed absolute accuracy metric drop (in original metrics value),
```
Comment (MaximProshin) [inline docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md:5]: Maybe a table will be even better here. Up to you.

File: docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md [hunk 10]
```diff
@@ -1,6 +1,17 @@
...
+
+3) `patience_epochs` - the number of epochs to train the model for a compression rate level set by the search algorithm before switching to another compression rate value.
+4) (Optional; default=0.025) `minimal_compression_rate_step` - The minimal compression rate change step value after which the training loop is terminated.
+5) (Optional; default=0.1) `initial_compression_rate_step` - Initial value for the compression rate increase/decrease training phase of the compression training loop.
+6) (Optional; default=0.5) `compression_rate_step_reduction_factor` - Factor used to reduce the compression rate change step in the adaptive compression training loop. 
+4) (Optional; default=1) `validate_every_n_epochs` - The parameter specifies across which number of epochs `Runner` should validate the compressed model.
+5) (Optional; default=1e4) `maximal_total_epochs` -  - number of training epochs, if the fine-tuning epoch reaches this number, the loop finishes the fine-tuning and return the model with the least accuracy drop.
```
Comment (MaximProshin) [inline docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md]: remove one dash

File: docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md [hunk 11]
```diff
@@ -1,6 +1,17 @@
...
+
 
 To launch the adaptive compression training loop, the user is expected to define several function related to model training, validation and optimizer creation (see [the usage documentation](../Usage.md#accuracy-aware-model-training) for more details) and pass them to the run method of an `AdaptiveCompressionTrainingLoop` instance. The training loop logic inside of the `AdaptiveCompressionTrainingLoop` is framework-agnostic, while all of the framework specifics are encapsulated inside of corresponding `Runner` objects, which are created and called inside the training loop. The adaptive compression training loop is generally aimed at automatically searching for the optimal compression rate in the model, with the parameters of the search algorithm specified in the configuration file as follows:
```
Comment (MaximProshin) [inline docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md:16]: the user is expected to define several function =>
the user should define several functions

File: docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md (hunk 7 above)
Comment (MaximProshin) [reply docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md]: I think it's confusing or I misunderstand it: can it be used for Pruning+INT8 for instance?

File: docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md [hunk 12]
```diff
@@ -28,6 +39,19 @@ To launch the adaptive compression training loop, the user is expected to define
...
+
+## Description of the work of Adaptive Compression Level training loop
+
+The first step is **Initial Training Phase** - It corresponds to the amount of epochs that the model is going to be trained for with the initial compression rate level/schedule set by the user in the standard NNCF manner (the initial pruning rate schedule above is an exponential schedule with the target pruning rate of 0.1).
+
+The second one is **Finding the optimal compression rate**, where the next compression rate value is determined by the search algorithm and the model is fine-tuned with that selected compression rate value for `"patience_epochs"` number of epochs. The process is continued until after the search algorithm terminates. The returned model is the model with the highest compression rate encountered during training given that is satisfies the accuracy drop criterion -- the accuracy value of the compressed model should not be more that `"maximal_relative_accuracy_degradation`" percent less that the original uncompressed model's accuracy value.
```
Comment (MaximProshin) [inline docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md:52]: The process is continued until after the search algorithm terminates =>
The process is continued until the search algorithm terminates

File: docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md [hunk 13]
```diff
@@ -28,6 +39,19 @@ To launch the adaptive compression training loop, the user is expected to define
...
+
+## Compression rate search algorithm
+
+The default behavior for the compression rate search algorithm implies changes in the compression rate level value by a step value that is decreasing throughout training.
+The training is terminated once the compression rate step value reaches the minimal value determined by the `"minimal_compression_rate_step"` parameter that can be specified in the `"params"` of `"accuracy_aware_training"` section.
+The initial value for the compression rate step is given be the `"initial_compression_rate_step"` parameter.
```
Comment (MaximProshin) [inline docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md:58]: is given be the => is given by the

File: docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md (hunk 12 above)
Comment (MaximProshin) [reply docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md:52]: The following is really hard to understand. Please rephrase it.
"the accuracy value of the compressed model should not be more that "maximal_relative_accuracy_degradation" percent less that the original uncompressed model's accuracy value"

File: docs/accuracy_aware_model_training/EarlyExitTrainig.md [hunk 14]
```diff
@@ -7,7 +7,9 @@ Note: since the EarlyExit training does not control any compression parameter th
 
 This training loop supports any combination of NNCF compression algorithms.
 
-There are only two parameters of Early-Exit training loop: `maximal_relative_accuracy_degradation` or `maximal_absolute_accuracy_degradation` - relative/absolute accuracy drop in percentage/in original metric with original, uncompressed model less than that is user tolerant. And `maximal_total_epochs` - number of training epochs, if the fine-tuning epoch reaches this number, the loop finishes the fine-tuning and return the model with the least accuracy drop
+There are only two main parameters of Early-Exit training loop: `maximal_relative_accuracy_degradation` or `maximal_absolute_accuracy_degradation` - relative/absolute accuracy drop in percentage/in original metric with original, uncompressed model less than that is user tolerant. And `maximal_total_epochs` - number of training epochs, if the fine-tuning epoch reaches this number, the loop finishes the fine-tuning and return the model with the least accuracy drop.
```
Comment (MaximProshin) [inline docs/accuracy_aware_model_training/EarlyExitTrainig.md:10]: return => returns

Comment (kshpv) [reply docs/accuracy_aware_model_training/EarlyExitTrainig.md]: Done

File: docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md (hunk 13 above)
Comment (kshpv) [reply docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md]: Done

File: docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md (hunk 12 above)
Comment (kshpv) [reply docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md]: Rephrased, please, take a look

File: docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md (hunk 11 above)
Comment (kshpv) [reply docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md]: Done

File: docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md (hunk 10 above)
Comment (kshpv) [reply docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md]: Done

File: docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md (hunk 9 above)
Comment (kshpv) [reply docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md:7]: Stayed with this one

File: docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md (hunk 8 above)
Comment (kshpv) [reply docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md]: Done

File: docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md (hunk 7 above)
Comment (kshpv) [reply docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md]: Update with examples. Please have a look

File: docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md (hunk 6 above)
Comment (kshpv) [reply docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md:1]: Added link in Usage.md

File: docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md (hunk 5 above)
Comment (kshpv) [reply docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md]: Done

File: docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md (hunk 4 above)
Comment (kshpv) [reply docs/accuracy_aware_model_training/AdaptiveCompressionLevelTraining.md]: Done

File: nncf/common/accuracy_aware_training/runner.py (hunk 3 above)
Comment (kshpv) [reply nncf/common/accuracy_aware_training/runner.py:177]: Are you satisfied?

Comment (AlexKoff88) [reply nncf/common/accuracy_aware_training/runner.py:177]: Ok, get it.

**Comment by kshpv** [comment]
Jenkins please retry a build

**Review by AlexKoff88 — APPROVED** [review]

**Review by alexsu52 — COMMENTED** [review]
Could you check that the accuracy-aware configs from the TensorFlow samples work well?

**Review by MaximProshin — APPROVED** [review]

**Review by alexsu52 — APPROVED** [review]
